[EDU-3202] feat: Fingerprint integration guide (#528)
aziontech/docs: src/content/docs/en/pages/guides/marketplace/integrations/fingerprint/fingerprint.mdx

---
title: How to install the Fingerprint integration through Azion Marketplace
description: Use this integration to capture devices and network data to better identify devices accessing your applications. 
meta_tags: protection, cybersecurity, edge computing
namespace: docs_use_case_fingerprint
permalink: /documentation/products/guides/fingerprint/
---

Azion **Fingerprint** enables you to capture devices and network data to better identify devices accessing your applications. It can be classified as a *Session Fingerprinting* integration and uses the collected data to generate a unique identifier and add it as a request header.

This integration includes two features:

> - A **browser-level script** (optional): a [JavaScript code](#embedding-the-script) that could be added to the web page. It captures data about the browser and sets it as a session cookie (named `_azfp_sc`), so this data is always sent by the browser. Even when the usage of this script is optional, it increases the precision of the integration.

Reviewer: ```suggestion
- A **browser-level script** (optional): a [JavaScript code](#embedding-the-script) that could be added to the web page. It captures data about the browser and sets it as a session cookie named `_azfp_sc`, so this data is always sent by the browser. Even when the usage of this script is optional, it increases the precision of the integration.
```

- An **edge function** (mandatory): a function to capture connection-related data as well as the "frontend fingerprint" (created by the browser script), when available, to create the session fingerprint. It'll then add this fingerprint as a request header (named `X-AZFP`).

---

## Embedding the script

> The browser script is a JavaScript code that runs in the web page you want to fingerprint, a process also know as JavaScript injection. It captures multiple data from the web browser:

Reviewer 2: ```suggestion
The browser script is a JavaScript code that runs on the web page you want to fingerprint, a process also known as JavaScript injection. It captures multiple data from the web browser:
```

- Audio and video support
- Canvas renderization
- If the browser has access to device-level API, such as accelerometer and gyroscope
- Manufacturer's data
- Timezone

> - WebGL renderization

Reviewer 2: ```suggestion
- WebGL rendering
```

Then, it creates a unique identifier containing all this information and sets it as a session cookie named `_azfp_sc`, as cookies are always sent in the requests made by the browser. The script checks if the session cookie already exists. If so, it won't create a new one.

Azion provides a code sample for JavaScript injection. 

1. You can find the code in [this file](https://mal2u8n8zk.map.azionedge.net/274addd29ac1edb2b200a297b98581cd88465e2ee36f18cfdbdc063402063fe5.js).
2. Then, add it on your HTML source. Example: 

```bash
<script src="anything.azion.xyz/arquivo.js"></script>
```

:::note
This is an optional step. However, using this script increases the precision of the **Fingerprint** integration. The code is a modularized script, in order to avoid collateral damage in any other scripts that might be running in the same web page.
:::

---

## Getting the integration

> To install Azion **Fingerptint**:

Reviewer 2: ```suggestion
To install Azion **Fingerprint**:
```

2. On the Marketplace's homepage, select the integration's card.
3. Once the integration's page opens, click the **Get It Now** button, at the bottom-right corner of the page.

A successful message appears to confirm your integration is installed.

:::tip
You can search any integration by browsing through the cards or typing a keyword in the search bar. 
:::

---

## Configuring the integration

### Setting up the Edge Firewall

To do so, follow the steps:

1. On the upper-left corner, select **Products menu** > **Edge Firewall** in the **SECURE** section.
2. Click the **Add Rule Set** button.
3. Give an easy-to-remember name to your new rule.
4. Select the domains you want to protect with the function.
5. Enable the **Edge Functions** switch in the **Edge Firewall Modules** section.
6. Click the **Save** button.

Done. Now you've instantiated the rule for your function and have access to edge functions on your edge firewall.

### Setting up the function

While still on the **Edge Firewall** page:

1. Select the **Functions** tab.
2. Click the **Add Function** button.
3. Give an easy-to-remember name to your instance.

> 4. On the dropdown menu, select the **Fingerptint** function.

Reviewer 2: ```suggestion
4. On the dropdown menu, select the **Fingerprint** function.
```

This action will load the function, showing a form with the function code and, just above it, two tabs: **Code** and **Args**.

> By clicking the Code tab, you'll be able to navigate through the source code but not to change it.

Reviewer 3: ```suggestion
By clicking the **Code** tab, you'll be able to navigate through the source code but not to change it.
```

The following data is used to create the fingerprint hash:

- Autonomous System Numbers (ASN)
- Cookie `_azfp_sc`
- Geoip City Code
- Geoip Country Code
- Geoip Region Code

> - Header Accept
> - Header Accept-CH
> - Header Accept-Encoding
> - Header Accept-Language
> - Header User-Agent
> - Header X-Forwarded-For

Reviewer: ```suggestion
- Header `Accept`
- Header `Accept-CH`
- Header `Accept-Encoding`
- Header `Accept-Language`
- Header `User-Agent`
- Header `X-Forwarded-For`
```

- Remote Address
- SSL Cipher
- SSL Version

On the **Args** tab, you'll pass your variables. In this case, you must declare just the variables that you want the function **to ignore**, as an **array of strings**. It's a not required action. If blank, the function will collect all the data.

```json
{
  "fingerprint_ignore": []
}
```

Possible variables to be ignored in the creation of the fingerprint:

- `cookie_fingerprint`: the cookie created by the [browser script](#embedding-the-script), named `_azfp_sc`
- `header_accept`: the content types the client can understand. Example: `text/html`

> - `header_accept_ch`: the client hints headers a client should include in subsequent requests. Exmaple: `Viewport-Width, Width`

Reviewer 2: ```suggestion
- `header_accept_ch`: the client hints headers a client should include in subsequent requests. Example: `Viewport-Width, Width`
```

- `header_accept_encoding`: the encoding algorithms a client supports. Example: `gzip`

> - `header_accept_language`: the language client prefers. Example: `"en-US"`

Reviewer 2: ```suggestion
- `header_accept_language`: the language the client prefers. Example: `"en-US"`
```

- `header_user_agent`: the software and version the client uses. Example: `Chrome/51.0.2704.106`
- `geoip_asn`: the ASN associated with the device's IP. Example: `52580`
- `geoip_city`: the city associated with the device's IP. Example: `"Brasilia"`
- `geoip_country_code`: the country code associated with the device's IP. Example: `"BR"`
- `geoip_region`: the region or state associated with the device's IP. Example: `"DF"`
- `remote_addr`: the device's remote IP address. Example: `000.0.0.0`
- `ssl_cipher`: the SSL cipher used in the device's connection. Example: `"ECDHE-RSA-AES128-GCM-SHA256"`
- `ssl_protocol`: the SSL protocol used in the device's connection. Example: `"TLSv1.2"`

:::note

> Any variable as "null" or blank will be still a valid value.

Reviewer: ```suggestion
Any variable as `null` or blank will be still a valid value.
```

:::

5. When you're done, click the **Save** button.

### Setting up the Rules Engine

To finish, you have to set up a **Rules Engine** to configure the *criteria* and the *behavior* to run the function.

Still in the **Edge Firewall** page:

1. Select the **Rules Engine** tab.
2. Click the **New Rule** button.
3. Give an easy-to-remember name to the rule.
   - You can add a description, but it's an optional step.
4. Select a *criteria* to run and catch the domains that you want to run the integration on.
    - Use this rule: 
5. Below, select a *behavior* to the *criteria*. In this case, it'll be **Run Function**.
6. Select the adequate function according to the name you gave it in the instantiate step.
7. Click the **Save** button.

> Done. You now have you can collect session data by using Azion **Fingerptint**.

Reviewer 2: ```suggestion
Done. You now have you can collect session data by using Azion **Fingerprint**.
```

---
title: How to install the Fingerprint integration through Azion Marketplace

> description: Use this integration to capture session data to better identify users and devices accessing your applications. 

Author: ```suggestion
description: Use this integration to capture session data to better identify devices accessing your applications. 
```

> meta_tags: bot, protection, cybersecurity, edge computing

Author: ```suggestion
meta_tags: protection, cybersecurity, edge computing
```

> description: The Azion Bot Protection analyzes incoming requests using edge functions on Edge Firewall, assigning a score based on rules and behaviors to detect and prevent malicious activities.

Author: ```suggestion
description: Use this integration to capture session data to better identify users and devices accessing your applications. 
```

> namespace: docs_use_case_bot_protection

Author: ```suggestion
namespace: docs_use_case_fingerprint
```

- An **edge function** (mandatory): a function to capture connection-related data as well as the "frontend fingerprint" (created by the browser script), when available, to create the session fingerprint. It'll then add this fingerprint as a request header (named `X-AZFP`).

---

## Embedding the script

The browser script is a JavaScript code that runs in the web page you want to fingerprint. It captures multiple data from the web browser:

- Audio and video support
- Canvas renderization
- If the browser has access to device-level API, such as accelerometer and gyroscope
- Manufacturer's data
- Timezone
- WebGL renderization

Then, it creates a unique identifier containing all this information and sets it as a session cookie named `_azfp_sc`, as cookies are always sent in the requests made by the browser. The script checks if the session cookie already exists. If so, it won't create a new one..

> You need to embed the whole code in the browser. To do so, proceed as follows:

Author: validate process

:::

---

## Configuring the integration

### Setting up the Edge Firewall

To do so, follow the steps:

1. On the upper-left corner, select **Products menu** > **Edge Firewall** in the **SECURE** section.
2. Click the **Add Rule Set** button.
3. Give an easy-to-remember name to your new rule.
4. Select the domains you want to protect with the function.
5. Enable the **Edge Functions** switch in the **Edge Firewall Modules** section.
6. Click the **Save** button.

Done. Now you've instantiated the rule for your function and have access to edge functions on your edge firewall.

### Setting up the function

While still on the **Edge Firewall** page:

1. Select the **Functions** tab.
2. Click the **Add Function** button.
3. Give an easy-to-remember name to your instance.
4. On the dropdown menu, select the **Fingerptint** function.

This action will load the function, showing a form with the function code and, just above it, two tabs: **Code** and **Args**.

By clicking the Code tab, you'll be able to navigate through the source code but not to change it.

The following data is used to create the fingerprint hash:

> - Autonomous System Numbers (ASN):
> - Cookie `_azfp_sc`: the cookie created by the [browser script](#embedding-the-script)
> - Geoip City Code: 
> - Geoip Country Code: 
> - Geoip Region Code: 
> - Header Accept: 
> - Header Accept-CH: 
> - Header Accept-Encoding: 
> - Header Accept-Language: 
> - Header User-Agent: 
> - Header X-Forwarded-For: 
> - Remote Address: 
> - SSL Cipher: 
> - SSL Version: 

Author: ```suggestion
- Autonomous System Numbers (ASN)
- Cookie `_azfp_sc`, the cookie created by the [browser script](#embedding-the-script)
- Geoip City Code
- Geoip Country Code
- Geoip Region Code
- Header Accept
- Header Accept-CH
- Header Accept-Encoding
- Header Accept-Language
- Header User-Agent
- Header X-Forwarded-For
- Remote Address
- SSL Cipher
- SSL Version
```

On the **Args** tab, you'll pass your variables. In this case, you must declare just the variables that you want the function **to ignore**, as an **array of strings**. It's a not required action. If blank, the function will collect all the data.

```json
{
  "fingerprint_ignore": []
}
```

Possible variables to be ignored in the creation of the fingerprint:

- `cookie_fingerprint`: the cookie created by the [browser script](#embedding-the-script), named `_azfp_sc`
- `header_accept`: the content types the client can understand. Example: `text/html`
- `header_accept_ch`: the client hints headers a client should include in subsequent requests. Exmaple: `Viewport-Width, Width`
- `header_accept_encoding`: the encoding algorithms a client supports. Example: `gzip`
- `header_accept_language`: the language client prefers. Example: `"en-US"`
- `header_user_agent`: the software and version the client uses. Example: `Chrome/51.0.2704.106`

> - `geoip_asn`: the ASN associated with the user's IP. Example: `52580`

Author: ```suggestion
- `geoip_asn`: the ASN associated with the device's IP. Example: `52580`
```
